ZEngine/src/Helper.cpp

```diff
@@ -0,0 +1,879 @@
+#include <pch.h>
+#include <ZEngineDef.h>
+#include <Helpers/BufferHelper.h>
+#include <Helpers/RendererHelper.h>
+#include <Helpers/MeshHelper.h>
+#include <Rendering/Renderers/Storages/IVertex.h>
+#include <Rendering/Shaders/Compilers/ShaderCompiler.h>
+
+#include <assimp/Importer.hpp>
+#include <assimp/postprocess.h>
+
+#include <Rendering/Buffers/FrameBuffers/FrameBufferSpecification.h>
+
+using namespace ZEngine::Rendering::Renderers;
+
+namespace ZEngine::Helpers
+{
+    void CreateBuffer(
+        Hardwares::VulkanDevice&                device,
+        VkBuffer&                               buffer,
+        VkDeviceMemory&                         device_memory,
+        VkDeviceSize                            byte_size,
+        VkBufferUsageFlags                      buffer_usage,
+        const VkPhysicalDeviceMemoryProperties& memory_properties,
+        VkMemoryPropertyFlags                   requested_properties)
+    {
+        VkBufferCreateInfo buffer_create_info = {};
+        buffer_create_info.sType              = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO;
+        buffer_create_info.size               = byte_size;
+        buffer_create_info.usage              = buffer_usage;
+        buffer_create_info.sharingMode        = VK_SHARING_MODE_EXCLUSIVE;
+
+        ZENGINE_VALIDATE_ASSERT(vkCreateBuffer(device.GetNativeDeviceHandle(), &buffer_create_info, nullptr, &buffer) == VK_SUCCESS, "Failed to create vertex buffer")
+
+        VkMemoryRequirements memory_requirement;
+        vkGetBufferMemoryRequirements(device.GetNativeDeviceHandle(), buffer, &memory_requirement);
+
+        bool     memory_type_index_found{false};
+        uint32_t memory_type_index{0};
+        for (; memory_type_index < memory_properties.memoryTypeCount; ++memory_type_index)
+        {
+            if ((memory_requirement.memoryTypeBits & (1 << memory_type_index)) &&
+                (memory_properties.memoryTypes[memory_type_index].propertyFlags & requested_properties) == requested_properties)
+            {
+                memory_type_index_found = true;
+                break;
+            }
+        }
+
+        if (!memory_type_index_found)
+        {
+            ZENGINE_CORE_ERROR("Failed to find a valid memory type")
+        }
+
+        VkMemoryAllocateInfo memory_allocate_info = {};
+        memory_allocate_info.sType                = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+        memory_allocate_info.allocationSize       = memory_requirement.size;
+        memory_allocate_info.memoryTypeIndex      = memory_type_index;
+
+        // ToDo : Call of vkAllocateMemory(...) isn't optimal because it is limited by maxMemoryAllocationCount
```

Comment:
Yes, what the comment says :)

The baseline maxMemoryAllocationCount the implementation has to guarantee to be Vulkan compliant is 1024 btw, so not that much all things considered.

Reply:
I created a issue to track it: https://github.com/JeanPhilippeKernel/RendererEngine/issues/221